### PR TITLE
hardware.rb: use `typed: strict`

### PR DESCRIPTION
**Pull request by p-linnane**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----
Part of https://github.com/Homebrew/brew/issues/17297.

### CODE REVIEW

File: Library/Homebrew/hardware.rb [hunk 1]
```diff
@@ -96,50 +98,59 @@ def family
         :dunno
       end
 
+      sig { returns(T.nilable(Integer)) }
```
Comment (Bo98) [inline Library/Homebrew/hardware.rb:101]:
```suggestion
      sig { returns(Integer) }
```

should hopefully work

File: Library/Homebrew/hardware.rb [hunk 2]
```diff
@@ -152,26 +163,32 @@ def arm?
...
       def big_endian?
         [1].pack("I") == [1].pack("N")
       end
 
+      sig { returns(FalseClass) }
       def virtualized?
         false
       end
 
+      sig { returns(T::Array[String]) }
```
Comment (Bo98) [inline Library/Homebrew/hardware.rb:181]:
```suggestion
      sig { returns(T::Array[Symbol]) }
```

the extend/os equivalents seem to be symbols

File: Library/Homebrew/hardware.rb [hunk 3]
```diff
@@ -152,26 +163,32 @@ def arm?
...
       def features
         []
       end
 
+      sig { params(name: T::Array[String]).returns(T::Boolean) }
```
Comment (Bo98) [inline Library/Homebrew/hardware.rb:186]:
```suggestion
      sig { params(name: Symbol).returns(T::Boolean) }
```

File: Library/Homebrew/hardware.rb [hunk 4]
```diff
@@ -186,6 +203,7 @@ def in_rosetta2?
   end
 
   class << self
+    sig { returns(T.nilable(T.any(Integer, String))) }
```
Comment (Bo98) [inline Library/Homebrew/hardware.rb:206]:
```suggestion
    sig { returns(String) }
```

If you also change the `else` clause to be `Hardware::CPU.cores.to_s` which would makes sense to me.
(can't include that in the suggestion as it's outside the diff range)

**Review by Bo98 — APPROVED** [review]

**Review by carlocab — APPROVED** [review]

**Comment by MikeMcQuaid** [comment]
Thanks @p-linnane!

**Pull request merged**